README.md
# Information about ipcalc

This is a modern tool to assist in network address calculations for IPv4 and IPv6.
It acts both as a tool to output human readable information about a network or address,
as well as a tool suitable to be used by scripts or other programs.

It supports printing a summary about the provided network address, multiple
command line options per information to be printed, transparent IPv6 support,
and in addition it will use libGeoIP if available to provide geographic information.

The project started as a fork of the ipcalc tool in the Fedora distribution
but it has since then replaced the original tool.


# How to build

This project is built using Meson and Ninja. To build it simply type:
```
$ meson setup build
$ ninja -C build
```

## Build options

There are a few build-time options supported by ipcalc. You can learn about
them, and optionally reconfigure the project before building it, by running the
following command after `meson setup build`:
```
$ meson configure build
```

In order to access geo-information, ipcalc depends on the maxmind
database (libmaxminddb). Alternatively it can be built using the older
libgeoip (libgeoip), or with no geo-information whatsoever. Meson will
detect which supported geo-information libraries are available on your system,
if any, and automatically choose the best one. You can override this choice by
providing `-Duse_maxminddb=enabled` to `meson setup` or `meson configure` to
require libmaxminddb or `-Duse_maxminddb=disabled -Duse_geoip=enabled` to
require libgeoip. See `meson configure build` for more options.

For example, the following command will compile with no support for
geo-information.
```
$ meson setup build `-Duse_maxminddb=disabled -Duse_geoip=disabled`
$ ninja -C build
```

## Optimizing the build

By default Meson optimizes for the most common type of build: those done by
developers. Consequently default builds of ipcalc with Meson are without
optimization and with debug information. If you want an optimized build,
pass `--buildtype=release` to `meson setup` or `meson configure`. See
[the Meson documentation](https://mesonbuild.com/Running-Meson.html#configuring-the-build-directory)
for more build options.
```
$ meson setup build --buildtype=release
$ ninja -C build
```


# Running unit tests

After the application has been configured and built for the host platform
(default), you may run the unit test suite against the ipcalc binary as
follows:
```
$ ninja -C build test
```


# Legacy build method

Although the Meson Build System, described above, is the recommended way to
build this application, the legacy GNU Makefile is still supported for older
systems that do not have the requisite versions of Meson and Ninja readily
available. The primary limitations of using the Makefile are that it does not
do automatic detection of libmaxminddb or libgeoip - you may have to specify
support for one or neither of those libraries manually - and there is no easy
way to run the test suite.

To build using the legacy GNU Makefile simply type:
```
$ make
```

In order to access geo-information the application depends on the maxmind
database (libmaxminddb). Alternatively it can be built using the older
libgeoip (libgeoip) or with no geo-information whatsoever. The options
can be provided on the makefile via the variables USE_GEOIP (yes/no),
USE_MAXMIND (yes/no). For example the following command will compile
with no support for geo-information.

```
$ make USE_GEOIP=no USE_MAXMIND=no
```


# Examples

## IPv4

```
$ ipcalc --all-info 193.92.150.2/24
Address:        193.92.150.2
Network:        193.92.150.0/24
Netmask:        255.255.255.0 = 24
Broadcast:      193.92.150.255
Reverse DNS:    150.92.193.in-addr.arpa.

Address space:  Internet
Address class:  Class C
HostMin:        193.92.150.1
HostMax:        193.92.150.254
Hosts/Net:      254

Country code:   GR
Country:        Greece
```

```
$ ipcalc -pnmb --minaddr --maxaddr --geoinfo --addrspace 193.92.150.2/255.255.255.224
NETMASK=255.255.255.224
PREFIX=27
BROADCAST=193.92.150.31
NETWORK=193.92.150.0
MINADDR=193.92.150.1
MAXADDR=193.92.150.30
ADDRSPACE="Internet"
COUNTRY="Greece"
```

## IPv6

```
$ ipcalc --all-info 2a03:2880:20:4f06:face:b00c:0:14/64
Full Address:   2a03:2880:0020:4f06:face:b00c:0000:0014
Address:        2a03:2880:20:4f06:face:b00c:0:14
Full Network:   2a03:2880:0020:4f06:0000:0000:0000:0000/64
Network:        2a03:2880:20:4f06::/64
Netmask:        ffff:ffff:ffff:ffff:: = 64
Reverse DNS:    6.0.f.4.0.2.0.0.0.8.8.2.3.0.a.2.ip6.arpa.

Address space:  Global Unicast
HostMin:        2a03:2880:20:4f06::
HostMax:        2a03:2880:20:4f06:ffff:ffff:ffff:ffff
Hosts/Net:      2^(64) = 18446744073709551616

Country code:   IE
Country:        Ireland
```

```
$ ipcalc -pnmb --minaddr --maxaddr --addrspace --geoinfo 2a03:2880:20:4f06:face:b00c:0:14/64
NETMASK=ffff:ffff:ffff:ffff::
PREFIX=64
NETWORK=2a03:2880:20:4f06::
MINADDR=2a03:2880:20:4f06::
MAXADDR=2a03:2880:20:4f06:ffff:ffff:ffff:ffff
ADDRSPACE="Global Unicast"
COUNTRY="Ireland"
```

## JSON output

```
$ ipcalc --all-info -j 2a03:2880:20:4f06:face:b00c:0:14/64
{
  "FULLADDRESS":"2a03:2880:0020:4f06:face:b00c:0000:0014",
  "ADDRESS":"2a03:2880:20:4f06:face:b00c:0:14",
  "FULLNETWORK":"2a03:2880:0020:4f06:0000:0000:0000:0000",
  "NETWORK":"2a03:2880:20:4f06::",
  "NETMASK":"ffff:ffff:ffff:ffff::",
  "PREFIX":"64",
  "REVERSEDNS":"6.0.f.4.0.2.0.0.0.8.8.2.3.0.a.2.ip6.arpa.",
  "ADDRSPACE":"Global Unicast",
  "MINADDR":"2a03:2880:20:4f06::",
  "MAXADDR":"2a03:2880:20:4f06:ffff:ffff:ffff:ffff",
  "ADDRESSES":"18446744073709551616",
  "COUNTRYCODE":"IE",
  "COUNTRY":"Ireland",
  "COORDINATES":"53.000000,-8.000000"
}
```


NEWS
* Version 1.0.3 (released 2023-06-07)
- When --no-decorate is given the default output will
  include no colors (#28)
- Correctly split networks with /31 (#25)


* Version 1.0.2 (released 2022-11-26)
- Fix ULA prefix generator to use only defined ULA range
- Corrected manpage generation


* Version 1.0.1 (released 2021-06-06)
- The application will now build even without ronn
- Improved JSON output on single host input


* Version 1.0.0 (released 2020-10-03)
- Manpage was converted to markdown


* Version 0.4.1 (released 2020-04-24)
- Fix identification of TEST-NET-1 and TEST-NET-2.
- Updated IANA IPv4 special purpose address registry to the
  2020-04-06 version.


* Version 0.4.0 (released 2019-12-28)
- Set the broadcast address according to RFC3021
- Added the --deaggregate (-d) option. This allows identifying the networks
  that cover a given IP range.


* Version 0.3.0 (released 2019-11-23)
- Added Meson build system; to build use:
   $ meson setup build
   $ ninja -C build
   $ ninja -C build test
- Extend --no-decorate support to work with the specific info output options.
- Added the -a/--address info output option to explicitly print the address.
- Do not display the address class information unless --all-info is given
  on command line.
- Fixed detection of IPv6 IETF Protocol Assignments address block
- Fixed detection of IPv6 6to4 address block
- Updated the IANA IPv6 Special-Purpose Address Registry support to
  2019-09-13, which adds a few new IPv6 address blocks that can now be
  detected by ipcalc.
- Added the -j/--json output mode to print the info using the JSON format.


* Version 0.2.5 (released 2019-02-18)
- Fixed compilation with no geoinformation
- Fixed segmentation fault when compiled with, but run without libmaxminddb.


* Version 0.2.4 (released 2018-07-23)
- Added support for libmaxmind DB API as an alternative to libgeoip.


* Version 0.2.3 (released 2018-05-26)
- Fixed --no-decorate option with --split and IPv6 addresses.
- Fixed build with USE_GEOIP=yes and USE_DYN_GEOIP=no (David Guglielmi)
- Fixed build with libgeoip < 1.6.3


* Version 0.2.2 (released 2017-12-27)
- Added a more silent version of --split when combined with
  --no-decorate option.


* Version 0.2.0 (released 2017-05-15)
- Added the --split (-S) option. This allows splitting networks
  into smaller chunks.


* Version 0.1.8 (released 2016-06-14)
- Do not print the reverse DNS information by default. Only
  when --all-info is specified.


* Version 0.1.7 (released 2016-04-01)
- Assume a classless IPv4 range by default. This breaks compatibility
  with previous versions since, unless --class-prefix is specified,
  the default IPv4 prefix is now /32.
- Added support for printing reverse DNS addresses.


* Version 0.1.6 (released 2015-12-08)
- Introduced color printing.


* Version 0.1.5 (released 2015-11-05)
- Auto-detect IPv6 addresses when generating random networks.
- Removed libpopt dependency by using getopt_long().


* Version 0.1.4 (released 2015-10-14)
- Added the --all-info option to enable the printing of all
  available information.


* Version 0.1.3 (released 2015-10-05)
- Print the country code in informational mode.


* Version 0.1.2 (released 2015-06-09)
- Print the country code when --geoinfo is requested.
- Only quote the strings if there are spaces in them.


* Version 0.1.1 (released 2015-05-19)
- Allow a prefix of 0, or 0.0.0.0 to be compatible with previous
  tool.


* Version 0.1.0 (released 2015-05-18)
- Added --random-private parameter to generate private networks.
- IPv4 and IPv6 addresses are now automatically being detected,
  and the --ipv4 or --ipv6 options are not mandatory to specify.
- Added the --minaddr and --maxaddr options to print the minimum
  and maximum addresses in the provided network.
- Added the --addresses option to print the number of addresses
  in the provided network.
- Added the --addrspace option to print the type of the network
  the address resides on.
- Added the --geoinfo option to print geographic information about
  the given IP address. That will work if libGeoIP is found on
  the system.
- By default, when no parameters are given, general information
  about the provided network/address are printed (also as --info
  option).



tests/ipcalc-testrunner.sh
#!/bin/sh
#
# This shell script is a simple test runner for ipcalc tests.
#
# Adapted from: Matej Susta <msusta@redhat.com>
#
#   Copyright (c) 2009 Red Hat, Inc. All rights reserved.
#
#   This copyrighted material is made available to anyone wishing
#   to use, modify, copy, or redistribute it subject to the terms
#   and conditions of the GNU General Public License version 2.
#
#   This program is distributed in the hope that it will be
#   useful, but WITHOUT ANY WARRANTY; without even the implied
#   warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#   PURPOSE. See the GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public
#   License along with this program; if not, write to the Free
#   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
#   Boston, MA 02110-1301, USA.

exitcode=0

ok() {
	echo "ok."
}

fail() {
	echo "FAILED!"
	exitcode=$((exitcode+1))
	echo -e "Output was:\n$1"
}

TestSuccessVerbatim() {
	echo -n "Checking $@... "
	output=$(sh -c '$1' 2>&1)
	rc=$?
	[ $rc -eq 0 ] && ok || fail $output
}

TestSuccess() {
	echo -n "Checking $@... "
	output=$(sh -c "$1" 2>&1)
	rc=$?
	[ $rc -eq 0 ] && ok || fail $output
}

TestFailure() {
	echo -n "Checking $@... "
	output=$(sh -c "$1" 2>&1)
	rc=$?
	[ $rc -eq 0 ] && fail $output || ok
}

TestOutput() {
	echo -n "Checking $1... "
	output=$(sh -c "$1" 2>&1)
	rc=$?
	[ "$output" = "$2" ] && ok || fail $output
}

TestOutputFile() {
	echo -n "Reading $2... "
	[ -e "$2" ] || { fail "missing file $2"; return; }
	contents="$(cat "$2" 2>/dev/null)"
	[ -n "$contents" ] && ok || { fail "failed to read $2"; return; }
	TestOutput "$1" "$contents"
}

TestEqual() {
	TestSuccess "$1"
	[ -n "$output" ] && output1="$output" || { fail "no output from $1"; return; }
	TestSuccess "$2"
	[ -n "$output" ] && output2="$output" || { fail "no output from $2"; return; }
	echo -n "Comparing output... "
	[ "$output1" = "$output2" ] && ok || fail "$output1 <> $output2"
}

while [ $# -gt 0 ]; do
	case $1 in
		--test-success-verbatim) TestSuccessVerbatim "$2"; shift ;;
		--test-success) TestSuccess "$2"; shift ;;
		--test-failure) TestFailure "$2"; shift ;;
		--test-output)  TestOutput "$2" "$3"; shift; shift ;;
		--test-outfile) TestOutputFile "$2" "$3"; shift; shift ;;
		--test-equal)   TestEqual "$2" "$3"; shift; shift ;;
		*) fail "invalid argument: $1" ;;
	esac
	shift
done

echo "$exitcode test(s) failed."
exit $exitcode


netsplit.c
/*
 * Copyright (c) 2003-2016  Simon Ekstrand
 * Copyright (c) 2010-2016  Joachim Nilsson
 * Copyright (c) 2016 Nikos Mavrogiannopoulos
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *  
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdint.h>

#include "ipv6.h"
#include "ipcalc.h"

static const char *numtoquad(uint32_t num)
{
	static char quad[64];

	num = htonl(num);
	return inet_ntop(AF_INET, &num, quad, sizeof(quad));
}

void show_split_networks_v4(unsigned split_prefix, const struct ip_info_st *info, unsigned flags)
{
	char buf[64];
	uint32_t diff, start, end;
	size_t maxlen = 0;
	unsigned count;
	uint32_t splitmask = ntohl(prefix2mask(split_prefix));
	uint32_t nmask = ntohl(prefix2mask(info->prefix));
	struct in_addr net, broadcast;
	unsigned jsonchain = JSON_FIRST;

	if (splitmask < nmask) {
		if (!beSilent)
			fprintf(stderr, "Cannot subnet to /%d with this base network, use a prefix > /%d\n",
				split_prefix, info->prefix);
		exit(1);
	}

	output_start(&jsonchain);
	array_start(&jsonchain, "Split networks", "SPLITNETWORK");

	if (inet_pton(AF_INET, info->network, &net) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv4 address: %s\n", info->network);
		exit(1);
	}
	net.s_addr = ntohl(net.s_addr);

	if (inet_pton(AF_INET, info->broadcast, &broadcast) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad broadcast address: %s\n", info->broadcast);
		exit(1);
	}
	broadcast.s_addr = ntohl(broadcast.s_addr);

	diff  = 0xffffffff - splitmask + 1;
	start = net.s_addr;
	end   = net.s_addr + diff - 1;

	/* Figure out max width of a network string. */
	while (1) {
		size_t len;

		len = snprintf(buf, sizeof(buf), "%s", numtoquad(start));
		if (len > maxlen)
			maxlen = len;

		start += diff;
		if (end == 0xffffffff || end >= broadcast.s_addr)
			break;
		end += diff;
	}

	start = net.s_addr;
	end = net.s_addr + diff - 1;
	count = 0;
	while (1) {
		if (!(flags & FLAG_NO_DECORATE) || (flags & FLAG_JSON)) {
			default_printf(&jsonchain, "Network:\t", NULL, "%s/%u", numtoquad(start), split_prefix);
		} else {
			printf("%s/%u\n", numtoquad(start), split_prefix);
		}

		count++;
		start += diff;
		if (end == 0xffffffff || end >= broadcast.s_addr)
			break;
		end += diff;
	}

	array_stop(&jsonchain);

	if ((!(flags & FLAG_NO_DECORATE)) || (flags & FLAG_JSON)) {
		dist_printf(&jsonchain, "\nTotal:  \t", "NETS", "%u", count);
		dist_printf(&jsonchain, "Hosts/Net:\t", "ADDRESSES", "%s", ipv4_prefix_to_hosts(buf, sizeof(buf), split_prefix));
	}

	output_stop(&jsonchain);
}

static const char *ipv6tostr(struct in6_addr *ip)
{
	static char str[64];

	return inet_ntop(AF_INET6, ip, str, sizeof(str));
}

void show_split_networks_v6(unsigned split_prefix, const struct ip_info_st *info, unsigned flags)
{
	int i, j, k;
	unsigned count;
	struct in6_addr splitmask, net, netmask, sdiff, ediff, start, end, tmpaddr, netlast;
	char buf[32];
	unsigned jsonchain = JSON_FIRST;

	if (inet_pton(AF_INET6, info->network, &net) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv6 address: %s\n", info->network);
		exit(1);
	}

	if (inet_pton(AF_INET6, info->hostmax, &netlast) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv6 address: %s\n", info->hostmax);
		exit(1);
	}

	if (inet_pton(AF_INET6, info->netmask, &netmask) <= 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: bad IPv6 mask: %s\n", info->netmask);
		exit(1);
	}

	if (ipv6_prefix_to_mask(split_prefix, &splitmask) < 0) {
		if (!beSilent)
			fprintf(stderr, "ipcalc: IPv6 prefix: %d\n", split_prefix);
		exit(1);
	}

	i = 0;
	j = 0;
	do {
		if (splitmask.s6_addr[i] > netmask.s6_addr[i])
			j = 1;
		if (netmask.s6_addr[i] > splitmask.s6_addr[i])
			j = 2;
		i++;
	} while (i < 16 && !j);

	if (j == 2) {
		if (!beSilent)
			fprintf(stderr, "Cannot subnet to /%d with this base network, use a prefix > /%d\n",
				split_prefix, info->prefix);
		exit(1);
	}

	memset(&sdiff, 0, sizeof(sdiff));
	memset(&ediff, 0, sizeof(ediff));

	for (i = 0; i < 16; i++) {
		sdiff.s6_addr[i] = 0xff - splitmask.s6_addr[i];
		end.s6_addr[i] = net.s6_addr[i] + sdiff.s6_addr[i];
	}

	memcpy(&start, &net, sizeof(net));
	memcpy(&ediff, &sdiff, sizeof(sdiff));

	memset(&tmpaddr, 0, sizeof(tmpaddr));
	tmpaddr.s6_addr[15] = 1;
	ipv6_add(&sdiff, &tmpaddr);

	output_start(&jsonchain);

	array_start(&jsonchain, "Split networks", "SPLITNETWORK");

	i = count = 0;
	while (!i) {
		if (!(flags & FLAG_NO_DECORATE) || (flags & FLAG_JSON)) {
			default_printf(&jsonchain, "Network:\t", NULL, "%s/%u", ipv6tostr(&start), split_prefix);
		} else {
			printf("%s/%u\n", ipv6tostr(&start), split_prefix);
		}

		ipv6_add(&start, &sdiff);

		j = 0;
		for (k = 0; k < 16; k+=2)
			if (end.s6_addr[k] != 0xff && end.s6_addr[k+1] != 0xff)
				j = 1;
		if (!j)
			i = 1;

		j = 0;
		k = 0;
		do {
			if (end.s6_addr[k] > netlast.s6_addr[k])
				j = 1;
			if (netlast.s6_addr[k] > end.s6_addr[k])
				j = 2;
			k++;
		} while (k < 16 && !j);

		if (!j || j == 1)
			i = 1;

		memset(&end, 0, sizeof(end));

		ipv6_add(&end, &start);
		ipv6_add(&end, &ediff);
		count++;
	}

	array_stop(&jsonchain);

	if ((!(flags & FLAG_NO_DECORATE)) || (flags & FLAG_JSON)) {
		dist_printf(&jsonchain, "\nTotal:  \t", "NETS", "%u", count);
		dist_printf(&jsonchain, "Hosts/Net:\t", "ADDRESSES", "%s", ipv6_prefix_to_hosts(buf, sizeof(buf), split_prefix));
	}

	output_stop(&jsonchain);

}



ipcalc-utils.c
/*
 * Copyright (c) 2018 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *   Martin Sehnoutka <msehnout@redhat.com>
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>

int __attribute__((__format__(printf, 2, 3))) safe_asprintf(char **strp, const char *fmt, ...)
{
	int ret;
	va_list args;

	va_start(args, fmt);
	ret = vasprintf(&(*strp), fmt, args);
	va_end(args);
	if (ret < 0) {
		fprintf(stderr, "Memory allocation failure\n");
		exit(1);
	}
	return ret;
}


ipcalc.h
/*
 * Copyright (c) 2016 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authors:
 *   Nikos Mavrogiannopoulos <nmav@redhat.com>
 */

#ifndef _IPCALC_H
#define _IPCALC_H

#if defined(USE_GEOIP)
  void geo_ip_lookup(const char *ip, char **country, char **ccode, char **city, char  **coord);
  int geo_setup(void);
# ifndef USE_RUNTIME_LINKING
#   define geo_setup() 0
# endif
#elif defined(USE_MAXMIND)
  void geo_ip_lookup(const char *ip, char **country, char **ccode, char **city, char  **coord);
  int geo_setup(void);
# ifndef USE_RUNTIME_LINKING
#   define geo_setup() 0
# endif
#else
# define geo_ipv4_lookup(x,y,z,w,a)
# define geo_ipv6_lookup(x,y,z,w,a)
# define geo_setup() -1
#endif

int __attribute__((__format__(printf, 2, 3))) safe_asprintf(char **strp, const char *fmt, ...);
char __attribute__((warn_unused_result)) *safe_strdup(const char *str);
int safe_atoi(const char *s, int *ret_i);

char *calc_reverse_dns4(struct in_addr ip, unsigned prefix, struct in_addr net, struct in_addr bcast);
char *calc_reverse_dns6(struct in6_addr *ip, unsigned prefix);

uint32_t prefix2mask(int prefix);
int ipv6_prefix_to_mask(unsigned prefix, struct in6_addr *mask);

struct in_addr calc_network(struct in_addr addr, int prefix);

char *ipv4_prefix_to_hosts(char *hosts, unsigned hosts_size, unsigned prefix);
char *ipv6_prefix_to_hosts(char *hosts, unsigned hosts_size, unsigned prefix);

typedef struct ip_info_st {
	char *ip;
	char *expanded_ip;
	char *expanded_network;
	char *reverse_dns;

	char *network;
	char *broadcast;	/* ipv4 only */
	char *netmask;
	char *hostname;
	char *geoip_country;
	char *geoip_ccode;
	char *geoip_city;
	char *geoip_coord;
	char hosts[64];		/* number of hosts in text */
	unsigned prefix;

	char *hostmin;
	char *hostmax;
	const char *type;
	const char *class;
} ip_info_st;

enum app_t {
	APP_VERSION=1,
	APP_CHECK_ADDRESS=1<<1,
	APP_SHOW_INFO=1<<2,
	APP_SPLIT=1<<3,
	APP_DEAGGREGATE=1<<4
};

#define FLAG_IPV6 (1<<1)
#define FLAG_IPV4 (1<<2)
#define FLAG_SHOW_MODERN_INFO (1<<3)
#define FLAG_RESOLVE_IP (1<<4)
#define FLAG_RESOLVE_HOST (1<<5)
#define FLAG_SHOW_BROADCAST (1<<6)
#define FLAG_SHOW_NETMASK (1<<7)
#define FLAG_SHOW_NETWORK (1<<8)
#define FLAG_SHOW_PREFIX (1<<9)
#define FLAG_SHOW_MINADDR (1<<10)
#define FLAG_SHOW_MAXADDR (1<<11)
#define FLAG_SHOW_ADDRESSES (1<<12)
#define FLAG_SHOW_ADDRSPACE (1<<13)
#define FLAG_GET_GEOIP (1<<14)
#define FLAG_SHOW_GEOIP ((1<<15)|FLAG_GET_GEOIP)
#define FLAG_SHOW_ALL_INFO (1<<16)
#define FLAG_SHOW_REVERSE (1<<17)
#define FLAG_ASSUME_CLASS_PREFIX (1<<18)
#define FLAG_NO_DECORATE (1<<20)
#define FLAG_SHOW_ADDRESS (1<<21)
#define FLAG_JSON (1<<22)
#define FLAG_RANDOM (1<<23)

/* Flags that are modifying an existing option */
#define FLAGS_TO_IGNORE (FLAG_IPV6|FLAG_IPV4|FLAG_GET_GEOIP|FLAG_NO_DECORATE|FLAG_JSON|FLAG_ASSUME_CLASS_PREFIX|(1<<16)|FLAG_RANDOM)
#define FLAGS_TO_IGNORE_MASK (~FLAGS_TO_IGNORE)

#define ENV_INFO_FLAGS (FLAG_SHOW_NETMASK|FLAG_SHOW_BROADCAST|FLAG_RESOLVE_IP|FLAG_RESOLVE_HOST|FLAG_SHOW_ADDRESS|FLAG_SHOW_REVERSE|FLAG_SHOW_GEOIP|FLAG_SHOW_ADDRSPACE|FLAG_SHOW_ADDRESSES|FLAG_SHOW_MAXADDR|FLAG_SHOW_MINADDR|FLAG_SHOW_PREFIX|FLAG_SHOW_NETWORK)
#define ENV_INFO_MASK (~ENV_INFO_FLAGS)

void show_split_networks_v4(unsigned split_prefix, const struct ip_info_st *info, unsigned flags);
void show_split_networks_v6(unsigned split_prefix, const struct ip_info_st *info, unsigned flags);

void deaggregate(char *str, unsigned flags);

#define KBLUE  "\x1B[34m"
#define KMAG   "\x1B[35m"
#define KRESET "\033[0m"

#define JSON_FIRST 0
#define JSON_NEXT  1
#define JSON_ARRAY_FIRST 2
#define JSON_ARRAY_NEXT  3

void
__attribute__ ((format(printf, 3, 4)))
color_printf(const char *color, const char *title, const char *fmt, ...);
void
__attribute__ ((format(printf, 3, 4)))
json_printf(unsigned * const jsonfirst, const char *jsontitle, const char *fmt, ...);
void va_color_printf(const char *color, const char *title, const char *fmt, va_list varglist);
void va_json_printf(unsigned  * const jsonfirst, const char *jsontitle, const char *fmt, va_list varglist);

void
__attribute__ ((format(printf, 4, 5)))
default_printf(unsigned * const jsonfirst, const char *title, const char *jsontitle, const char *fmt, ...);
void
__attribute__ ((format(printf, 4, 5)))
dist_printf(unsigned * const jsonfirst, const char *title, const char *jsontitle, const char *fmt, ...);

void array_start(unsigned * const jsonfirst, const char *head, const char *json_head);
void array_stop(unsigned * const jsonfirst);
void output_start(unsigned * const jsonfirst);
void output_stop(unsigned * const jsonfirst);

extern int beSilent;

#endif
